os/arch/arm/src/s5j/s5j_ticksuppress.c

```diff
@@ -0,0 +1,273 @@
+		case READ:
+			return getreg32(S5J_RTC_CURTICCNT0);
+		case WRITE:
+			putreg32(SYSTICK_RELOAD * value, S5J_RTC_TICCNT0);
+			break;
+		case STOP:
+			/* Clear interrupt pending */
+
+			putreg32(RTC_INTP_TIMETIC0, S5J_RTC_INTP);
+
+			/* Disable timertick */
+
+			modifyreg32(S5J_RTC_RTCCON, RTC_RTCCON_TICKEN0_ENABLE, 0);
+			break;
+		default:
+			/* Shouldn't reach here */
+			DEBUGPANIC();
+			break;
+	}
+
+	return 0;
+}
+
+static void s5j_set_next_event(uint32_t next_event)
+{
+	/* Disable timer tick */
+
+	s5j_timer_setup(STOP, 0);
+
+	/* Configure the RTC timetick to generate interrupt on next event */
+
+	s5j_timer_setup(WRITE, next_event);
+
+	/* Enable timer tick */
+
+	s5j_timer_setup(START, 0);
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Function:  up_stop_ticksuppress
+ *
+ * Description:
+ *   Disables tick suppression (re-enables periodic ticks) when,
+ *   1. Scheduled event expires
+ *   2. Context switch takes place on an external event (interrupt)
+ *
+ ****************************************************************************/
+
+void up_stop_ticksuppress(void)
+{
+	/* If tick suppression is disabled, do nothing! */
+
+	if (g_tick_mode == TICKSUPPRESS_DISABLE) {
+		return;
+	}
+
+	/* If this function called from interrupt context,
+	 * read current value of the counter and update last event value
+	 */
+
+	if (up_interrupt_context()) {
+		last_event = ((uint32_t)s5j_timer_setup(READ, 0) / SYSTICK_RELOAD);
+	}
+
+	/* This function will be called when last scheduled event expires.
+	 * so update the global system timer using expired last event value.
+	 */
+
+	if (last_event) {
+		/* Update global system timer */
+
+		g_system_timer += last_event;
+
+		/* Update wdog delay */
+
+		wd_updatedelay(last_event);
+
+		/* Reset last event */
+
+		last_event = 0;
+	}
+
+	/* Change timer tick state as periodic */
+
+	g_tick_mode = TICKSUPPRESS_DISABLE;
+
+	/* Re-enable periodic timer ticks */
+
+	s5j_set_next_event(0x1);
+}
+
+/*****************************************************************************
+ * Function:  up_start_ticksuppress
+ *
+ * Description:
+ *   Enables tick suppression (disables periodic ticks), when system goes idle
+ *
+ *****************************************************************************/
+
+void up_start_ticksuppress(void)
+{
+	uint32_t next_event = 0;
+
+	/* Check whether the threshold value is reached or not
+	 * If threshold value is crossed, ticks are suppressed and
+	 * next expected event is scheduled.
+	 */
+
+	s5j_timer_setup(STOP, 0);
+
+	next_event = s5j_get_next_event();
+	if (!next_event) {
+		next_event = MAX_TCNT;
+	}
+	last_event = next_event;
+
+	g_tick_mode = TICKSUPPRESS_ENABLE;
+
+	/* Schedule next expected event */
+
+	s5j_set_next_event(next_event);
```

Comment:
As i observe, to change the timer to periodic, we are passing 0x1 as argument as below 
s5j_set_next_event(0x1); 
What if the next_event returned by s5j_get_next_event() is also 0x1 ? what is expected behavior?

Reply:
timer interrupt comes after 1 tick